gordon/metrics/log.py

```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+# TODO: update plugin loader to load metrics logger as default
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.MetricLogger): Object used for outputting metrics.
+        time_unit (number): (optional) Multiplier used to scale time.time()
+            output.
+    """
+    def __init__(self, metric, logger, time_unit=1):
+        self.metric = metric
+        self.logger = logger
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer.
+        """
+        self._start_time = time.time()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger.
+        """
+        time_elapsed = (time.time() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class MetricLogger:
```

Comment:
Is there internal magic that piggy-backs the regular logger config, i.e. sends to stream/syslog/stackdriver?

Actually, how come this isn't itself a logger (i.e. `logging.Logger`)? and registers itself as a gordon metrics logger?

Reply:
> Is there internal magic that piggy-backs the regular logger config, i.e. sends to stream/syslog/stackdriver?

Yeah, I think that `ulogger` sets everything up so that further calls to eg. `logging.info` make use of that set up, like it happens in [gordon/main.py](https://github.com/spotify/gordon/blob/develop/gordon/main.py#L203).

> Actually, how come this isn't itself a logger (i.e. logging.Logger)?

I don't see the benefits for the metrics logger to log to a separate logger. I don't have a strong opinion on this, so I'm curious if you think that would be a better approach?

Reply:
The benefit would be to hook into the dynamic log level changes that the introspection server plugin can give; i.e. we can turn on metrics logging even if we have ffwd (not sure if that is even possible to have two metrics relays - though an interesting idea). But more so just to toggle off/on, and to change its level. @descent3me and I even entertained the idea of having "log level"-like assignments to metric types (maybe a hack day project?).

It also seems weird to me that it's behaving like a logger, but not really as it only just makes use of `logging.<level>(msg)` and essentially attaches itself to the service logger (basically, I like separation of responsibilities). A logger gives us the ability to format how we want the message line to be ([as you're familiar with](https://github.com/spotify/ulogger/blob/master/ulogger/syslog.py#L107)), [add context](https://docs.python.org/3/howto/logging-cookbook.html#adding-contextual-information-to-your-logging-output), [add record attributes](https://docs.python.org/3/howto/logging-cookbook.html#customizing-logrecord), and [allow users to filter](https://docs.python.org/3/library/logging.html#filter-objects).

Reply:
@econchick Gotcha! I still have some questions, but let's talk IRL tomorrow.